### DISCUSSION TITLE
change texture image transformation properties

### DISCUSSION THREAD

**Discussion by orenmiron** [google/model-viewer#3255] [Q&A]
I am trying to rotate a texture that is already attached to a mesh in the scene.
i can access the texture, by using the code below. but i couldn't find any rotation property or matrix.
please advise.

code:
let t = modelViewer.model.materials[0].pbrMetallicRoughness.baseColorTexture.texture;

**Comment by elalish** [comment]
Our API doesn't expose changing that glTF extension property yet, so you'll have to do it where the glTF is authored. What is your use case?

**Reply by orenmiron** [reply]
Thank you for the rapid response.
I am using model-viewer for gltf files created by my application.
My gltf files use the KHR_texture_transform extension. with non-uniform scale and rotate.
see attached file
 ThreeJS implementation order is "rotate / scale" and other 3D viewers use "scale / rotate".
I would like to prevent the user viewing different textures on different viewers.
rotating the UV is not possible because different channels on the same material have different rotation angles.
( I saw https://github.com/google/model-viewer/issues/3240, but maybe there is another way to fix my problem).
[TankTop_Front.zip](https://github.com/google/model-viewer/files/8226998/TankTop_Front.zip)

**Reply by elalish** [reply]
Hmm, your file seems to be malformed and doesn't load. Can you try packing it into a GLB? Can you show some screenshots to describe what you mean by "rotate / scale" and "scale / rotate"? This should all be well-defined in the glTF spec, so if you're getting different results, it may be a bug in one or the other. 

I still can't tell though; are you actually trying to rotate a texture after loading for some reason, or is the loaded model just incorrect?

**Reply by elalish** [reply]
Check #3263 to see if this fix helps you.

**Reply by orenmiron** [reply]
Sorry, I couldn't convert my gltf to glb without loosing my texture transformations.
I am not replacing my textures on runtime. just loading my gltf for viewing. no editing.
Try again the new zip file, I moved the textures to the gltf folder.
and i added snapshots.
model-viewer is on the left side, Microsoft Windows 10 3D viewer is on the right side.
![snapshot](https://user-images.githubusercontent.com/7829843/157865370-5c39d9c8-00a5-437a-a1d2-0b758fb761c0.jpg)
![snapshot2](https://user-images.githubusercontent.com/7829843/157865888-bcd5a31e-893c-456f-a40b-08ad40ece8cd.jpg)

[TankTop_Front.zip](https://github.com/google/model-viewer/files/8231742/TankTop_Front.zip)

**Reply by elalish** [reply]
The fix hasn't been released yet. Can you try replacing textures in our editor and see if it's working right there? You can also build locally to get the latest version on master for testing.

**Comment by IvanViale** [comment]
Hi evryone,
did you find a solution to change offset and scale of the texture parameters?
I also have a gltf file with the KHR_texture_transform extension and i need to adapt the same texture to different part of my object, there is maybe another solution?
Can I access to this parameters programmatically?
Thanks

![image](https://user-images.githubusercontent.com/36745926/221422850-d1e0a048-0c26-4bf4-9c31-fabf9359b9f8.png)

**Reply by elalish** [reply]
@mrdoob I think this is related to our earlier discussion about how multiple UV sets behave. I'm thinking I'll be able to expose this API once your changes are finalized, right?

**Reply by mrdoob** [reply]
Yep!

**Reply by IvanViale** [reply]
When you can expose this API?

**Reply by elalish** [reply]
Not sure when I'll get to it. If the next version of three.js has support for this, you're welcome to make a PR then.
